crates/core/component/dex/src/component/position_manager/counter.rs

```diff
@@ -24,13 +25,45 @@ pub(crate) trait PositionCounter: StateWrite {
             return 0;
         };
 
-        // This is safe because we only increment the counter via a [`Self::increase_position_counter`].
+        // This is safe because we only increment the counter via [`Self::increase_position_counter`].
         let raw_count: [u8; 2] = raw_count
             .try_into()
             .expect("position counter is at most two bytes");
         u16::from_be_bytes(raw_count)
     }
+}
 
+impl<T: StateRead + ?Sized> PositionCounterRead for T {}
+
+#[async_trait]
+pub(crate) trait PositionCounter: StateWrite {
+    async fn update_trading_pair_position_counter(
+        &mut self,
+        prev_state: &Option<Position>,
+        new_state: &Position,
+        _id: &position::Id,
+    ) -> Result<()> {
+        use position::State::*;
+        let trading_pair = new_state.phi.pair;
+
+        match prev_state {
+            Some(prev_state) => match (prev_state.state, new_state.state) {
+                (Opened, Closed) => {
+                    let _ = self.decrement_position_counter(&trading_pair).await?;
+                }
+                _ => {}
+            },
+
+            None => {
+                let _ = self.increment_position_counter(&trading_pair).await?;
+            }
+        }
```

Comment:
IMO this match statement would be clearer if there was a bit of prep work with a map clause, like
```rust
match (prev_state.map(|p| p.state), new_state.state) {
    // Increment the counter whenever a new position is opened
    (None, Opened) => self.increment_position_counter(&trading_pair).await?,
    // Decrement the counter whenever an opened position is closed
    (Some(Opened), Closed) => self.decrement_position_counter(&trading_pair).await?,
    // Other state transitions don't affect the opened position counter
    _ => {}
}
```